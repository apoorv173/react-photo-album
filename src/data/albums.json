[
  {
    "username": 1,
    "albumId": 1,
    "albumTitle": "quidem molestiae enim",
    "color": "fff"
  },
  {
    "username": 1,
    "albumId": 2,
    "albumTitle": "sunt qui excepturi placeat culpa",
    "color": "111"
  }
]