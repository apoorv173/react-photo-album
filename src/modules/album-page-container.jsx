import React, { Suspense, lazy, useEffect } from 'react';
import { connect } from 'react-redux';
import queryString from 'query-string';
import { Loader, Error } from '../components';
import { fetchAllAlbums } from '../actions/album-actions';
import { getAlbumDetails } from '../selectors/selectors';

const AlbumList = lazy(() => import("../components/album-list/album-list"));
const DataTable = lazy(() => import("../components/data-table/data-table"));
const Pagination = lazy(() => import("../components/pagination/pagination"));

const AlbumPage = ({ albumsData, fetchAlbums, history, match, loading, error }) => {

	const getQueryParams = queryString.parse(history.location.search);
	const getSizeFromURL = parseInt(getQueryParams["size"], 10) || 20;
	const getPageFromURL = parseInt(getQueryParams["page"]) || 1;

	const albumCount = parseInt((albumsData ? albumsData.length : 0), 10);
	
	useEffect(() => {
		fetchAlbums(getPageFromURL, getSizeFromURL);
	}, [match]);
	
	const onCountChange = (value) => {
		history.push({
			pathname: `/albums`,
			search: `?size=${value}&page=1`
		});
	}

	const onPrevClick = () => {
		if(getPageFromURL > 1) {
			const page = getPageFromURL-1;
			history.push({
				search: `?size=${getSizeFromURL}&page=${page}`
			});
		}
	}

	const onNextClick = () => {
		const page = getPageFromURL+1;
		history.push({
			search: `?size=${getSizeFromURL}&page=${page}`
		});
	}
	switch(true) {
		case (error && error !== ''):
			return <Error error={error} />;

		case (albumsData && albumsData.length >= 0):
			return (
				<div className="">
					<h1>All albums</h1>
					<Suspense fallback={ <Loader />} >
						<Pagination 
							prevDisabled={(getPageFromURL <= 1)}
							nextDisabled={albumCount < getSizeFromURL}
							onPrevClick={onPrevClick}
							onNextClick={onNextClick}
						/>
						<DataTable 
							rowCount={getSizeFromURL}
							onCountChange={onCountChange}
						/>
						<AlbumList 
							albums={albumsData}
						/>
					</Suspense>
				</div>
			);
		default: 
			return null;
	}
};

const mapStateToProps = state => {
    return {
		albumsData: getAlbumDetails(state),
		error: state.albums.error,
		loading: state.albums.loading
    }
};

const mapDispatchToProps = dispatch => {
    return {
        fetchAlbums: (size, limit) => {
          dispatch(fetchAllAlbums(size, limit));
        }
      }
};

export default connect(mapStateToProps, mapDispatchToProps)(AlbumPage);
