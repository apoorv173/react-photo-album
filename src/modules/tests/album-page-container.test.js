import React, { lazy } from "react";
import {MemoryRouter} from "react-router-dom";
import { shallow, mount } from "enzyme";
import { AlbumPage } from "../album-page-container";
import albums from '../../data/more-albums.json';
import { history } from "../../configureStore";

const AlbumList = lazy(() => import("../components/album-list/album-list"));
const DataTable = lazy(() => import("../components/data-table/data-table"));
const Pagination = lazy(() => import("../../components/pagination/pagination"));

describe("Album List Component", () => {
    it("Should return null if albums doesnt exist, and no eror is there", () => {
        const tree = shallow(<MemoryRouter><AlbumPage loading={false} error="" albumsData={[]} /></MemoryRouter>)
        expect(tree).toMatchSnapshot();
    });
    
    it('rendered lazily on click of the next button', async()=> {
        const fetchAlbums = jest.fn();
        const wrapper = mount(
            <MemoryRouter>
                <AlbumPage  albumsData={albums} loading={false} error='' history={history} fetchAlbums={fetchAlbums} >
                </AlbumPage>
            </MemoryRouter>
        );
    
        await AlbumList;
        await DataTable;
        await Pagination;

        wrapper.setProps({ albumsData: albums, loading: false });
        wrapper.find('.next-btn').simulate('click');
        expect(history.location.search).toEqual('?size=20&page=2');
      });

      it('rendered lazily on click of the previous button', async()=> {
        const fetchAlbums = jest.fn();
        const wrapper = mount(
            <MemoryRouter>
                <AlbumPage  albumsData={albums} loading={false} error='' history={history} fetchAlbums={fetchAlbums} >
                </AlbumPage>
            </MemoryRouter>
        );
    
        wrapper.find('.prev-btn').simulate('click');
        expect(history.location.search).toEqual('?size=20&page=1');
        wrapper.unmount();
      });

    it("Should render the error component only with its snapshot", () => {
        const fetchAlbums = jest.fn();
        const wrapper = mount(
            <MemoryRouter>
                <AlbumPage  albumsData={[]} loading={false} error='No results found' history={history} fetchAlbums={fetchAlbums} >
                </AlbumPage>
            </MemoryRouter>
        );
        expect(wrapper.find('h1').length).toBe(1);
        expect(wrapper.find('h1').text()).toBe('No results found');
        wrapper.unmount();
    });

    it("Should render the loader component only with its snapshot", () => {
        const fetchAlbums = jest.fn();
        const wrapper = mount(
            <MemoryRouter>
                <AlbumPage  albumsData={[]} loading={true} error='' history={history} fetchAlbums={fetchAlbums} >
                </AlbumPage>
            </MemoryRouter>
        );
        expect(wrapper.find('.loader').length).toBe(1);
        wrapper.unmount();
    });

});