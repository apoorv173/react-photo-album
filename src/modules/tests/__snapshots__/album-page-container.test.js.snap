// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album List Component Should render the error component only with its snapshot 1`] = `ShallowWrapper {}`;

exports[`Album List Component Should render the loader component only with its snapshot 1`] = `ShallowWrapper {}`;

exports[`Album List Component Should return null if albums doesnt exist, and no eror is there 1`] = `ShallowWrapper {}`;

exports[`Album List Component rendered lazily 1`] = `
<div>
  <h1>
    All albums
  </h1>
  <nav
    aria-label="Page navigation"
  >
    <ul
      className="pagination"
    >
      <li
        className="left"
      >
        <button
          aria-label="Previous"
          className="btn btn-primary prev-btn"
          disabled="disabled"
          onClick={[Function]}
        >
          <span
            aria-hidden="true"
          >
            Previous
          </span>
        </button>
      </li>
      <li
        className="right"
      >
        <button
          aria-label="Next"
          className="btn btn-primary next-btn"
          disabled="disabled"
          onClick={[Function]}
        >
          <span
            aria-hidden="true"
          >
            Next
          </span>
        </button>
      </li>
    </ul>
  </nav>
  <nav
    aria-label="Page rows"
  >
    <h4
      className="d-inline-block"
    >
      Rows per page: 
    </h4>
    <span>
          
    </span>
    <select
      className="form-control select-wrapper d-inline-block"
      onChange={[Function]}
      value={20}
    >
      <option
        value={20}
      >
        20
      </option>
      <option
        value={30}
      >
        30
      </option>
      <option
        value={50}
      >
        50
      </option>
    </select>
  </nav>
  <div
    className="row"
  >
    <a
      href="/albums/1/photos"
      onClick={[Function]}
    >
      <div
        className="thumbnail-wrapper col-xs-12 col-sm-3"
      >
        <div
          className="thumbnail"
        >
          <div
            className="layer"
          >
            <h4
              className="thumbnail-heading"
            >
              quidem molestiae enim
            </h4>
            <p
              className="thumbnail-text"
            >
               By
                
              1
            </p>
          </div>
          <img
            className="thumbnail-img"
            src="https://via.placeholder.com/150/fff"
          />
        </div>
      </div>
    </a>
    <a
      href="/albums/2/photos"
      onClick={[Function]}
    >
      <div
        className="thumbnail-wrapper col-xs-12 col-sm-3"
      >
        <div
          className="thumbnail"
        >
          <div
            className="layer"
          >
            <h4
              className="thumbnail-heading"
            >
              sunt qui excepturi placeat culpa
            </h4>
            <p
              className="thumbnail-text"
            >
               By
                
              1
            </p>
          </div>
          <img
            className="thumbnail-img"
            src="https://via.placeholder.com/150/111"
          />
        </div>
      </div>
    </a>
  </div>
</div>
`;
